File: client/common/messages.go
package common

import (
	"log"
	"strings"
)

type SubmitMessage struct{ Payload []Bet }
type GetWinnersMessage struct{}
type SubmitResultMessage struct{ Payload string }
type WinnersMessage struct{ Payload []Dni }

type MessageType = string

const (
	// Client messages
	SUBMIT      MessageType = "S"
	GET_WINNERS MessageType = "G"

	// Server messages
	SUBMIT_RESULT MessageType = "R"
	WINNERS       MessageType = "W"

	SPLIT_CHAR = "\n"
)

func NewSubmitMessage(bets []Bet) *SubmitMessage {
	return &SubmitMessage{bets}
}

func (m *SubmitMessage) ToString() string {
	betsString := []string{}
	for i := 0; i < len(m.Payload); i++ {
		betsString = append(betsString, m.Payload[i].ToString())
	}
	return SUBMIT + strings.Join(betsString, SPLIT_CHAR)
}

func NewGetWinnersMessage() *GetWinnersMessage {
	return &GetWinnersMessage{}
}

func (m *GetWinnersMessage) ToString() string {
	return GET_WINNERS
}

func splitMessage(msg string, expectedType MessageType) string {
	msgType := msg[0:1]
	if msgType != expectedType {
		log.Fatalf("Expected message of type %s from server but got type %s", expectedType, msgType)
	}
	return msg[1:]
}

func WinnersMessageFromString(msg string) *WinnersMessage {
	dnis := splitMessage(msg, WINNERS)
	return &WinnersMessage{
		strings.Split(dnis, SPLIT_CHAR),
	}
}

func SubmitResultMessageFromString(msg string) *SubmitResultMessage {
	result := splitMessage(msg, SUBMIT_RESULT)
	return &SubmitResultMessage{
		result,
	}
}

File: server/common/messages.py
import string
from enum import Enum

from .utils import Bet

class MessageTypes(str, Enum):
    # Client messages
    SUBMIT = "S" # Payload: [Bet]
    GET_WINNERS = "G" # Payload: None

    # Server messages
    SUBMIT_RESULT = "R" # Payload: str
    WINNERS = "W" # Payload: [str]


SPLIT_CHAR = "\n"

class Message:
    def __init__(self, type: str, payload):
        self.type = type
        self.payload = payload

    def from_string(message_string: str) -> "Message":
        type = MessageTypes(message_string[0])
        rest = message_string[1:]
        if type == MessageTypes.SUBMIT:
            bets = [Bet.from_string(x) for x in rest.split(SPLIT_CHAR)]
            return Message(type, bets)
        elif type == MessageTypes.GET_WINNERS:
            return Message(type, rest)
        raise Exception("Unknown message type")

    def to_string(self) -> string:
        encoded = self.type
        if self.type == MessageTypes.SUBMIT_RESULT:
            encoded += self.payload
        elif self.type == MessageTypes.WINNERS:
            encoded += SPLIT_CHAR.join(self.payload)
        else:
            raise Exception("Not implemented")
        return encoded